draft-ietf-quic-transport.md

```diff
@@ -1037,8 +1037,18 @@ Servers MUST drop other packets that contain unsupported versions.
 
 Packets with a supported version, or no version field, are matched to a
 connection using the connection ID or - for packets with zero-length connection
-IDs - the address tuple.  If the packet doesn't match an existing connection,
-the server continues below.
+IDs - the address tuple, with the following exception.
+
+A server that uses a non-zero-length connection ID SHOULD handle Initial packets
+that share the same address tuple, Source and Destination Connection IDs, but
+contain different first ClientHello messages as belonging to different
+connections.  This might be used by a client to defend against attacks that race
+spoofed Initial packets with legitimate ones.  A server can treat every Initial
+packet containing a CRYPTO frame at an offset of 0 as potentially creating a new
+connection, discarding any packet that has the same Destination Connection ID
+and CRYPTO frame content as one that has been answered.
+
```

Comment:
Yeah the text that discusses the approach purely by referring to the CRYPTO frame is hard to understand (and it might be erroneous).

The logic I'd expect the stacks to have is something like below:
```
if crypto_frame.offset == 0:
  ch_hash = sha256(crypto_frame)
  if conn_is_open():
    if open_conn.handshake_key != nil:
      if open_conn.ch_hash != ch_hash:
        handle_as_new_connection()
      else:
        handle_packet_with_care()             // to be addressed in #2045, #2053
    else:
      handle_initial_as_part_of_connection()  // HRR case
  else:
    handle_as_new_connection()
```

> I am concerned that to make this work, we should go back to the "stateless HRR" model, maybe sending the TLS HRR in a crypto frame in the payload of a retry packet.

I do not think so, because our experience in TLS 1.3 tells us that HRR will rarely be used, because the endpoints are incentivized to avoid the additional round-trip imposed by HRR.